### PR TITLE
Handle getDetails JSON error on 5xx status code

### PR DESCRIPTION
**Pull request by PriOliveira**
Should solve https://github.com/ipinfo/python/issues/107

### CODE REVIEW

**Review by max-ipinfo — CHANGES_REQUESTED** [review]
Thank you @PriOliveira for filing #107 and for sending a PR for this change! Your solution definitely improves the robustness of our library. 

I would like to generalize and improve your solution further by explicitly handling non-JSON responses. Could you verify that it works on your end?

File: ipinfo/handler.py [hunk 1]
```diff
@@ -120,9 +120,12 @@ def getDetails(self, ip_address=None, timeout=None):
             url += "/" + ip_address
         headers = handler_utils.get_headers(self.access_token, self.headers)
         response = requests.get(url, headers=headers, **req_opts)
-        if response.status_code == 429:
-            raise RequestQuotaExceededError()
-        if response.status_code >= 400:
+        if response.status_code >= 500:
```
Comment (max-ipinfo) [inline ipinfo/handler.py:123]: The only issue with this approach is that we lose the error details when `response.status_code >= 500`.

A better approach is to explicitly handle the response content type and extract the error message.

Could you change the lines (and the same in `handler_async.py`) to:
```
if response.status_code == 429:
    raise RequestQuotaExceededError()
if response.status_code >= 400:
    content_type = response.headers.get('Content-Type')
    error_code = response.status_code
    if content_type == 'application/json':
        error_response = response.json()
    else:
        error_response = {'error': response.text}
    raise APIError(error_code, json.dumps(error_response))
```

Thanks!

Comment (PriOliveira) [reply ipinfo/handler.py:123]: Sure, good point. I'll update.

File: ipinfo/handler.py [hunk 2]
```diff
@@ -123,9 +123,15 @@ def getDetails(self, ip_address=None, timeout=None):
         if response.status_code == 429:
             raise RequestQuotaExceededError()
         if response.status_code >= 400:
-            error_response = response.json()
             error_code = response.status_code
+
+            content_type = response.headers.get('content-type')
```
Comment (max-ipinfo) [inline ipinfo/handler.py:128]: Two final formatting/linting nitpicks:
- remove blank lines 127 and 134 (and lines 149 and 156 in `handler_async.py`)
- even though HTTP headers are case-insensitive, we prefer to use the capitalized form in our code

Comment (PriOliveira) [reply ipinfo/handler.py:128]: For the header, I was trying to follow the existing pattern on those files. I only saw examples with lower case there.

**Comment by PriOliveira** [comment]
I added some tests to try to cover those cases, hope you don't mind.

File: ipinfo/handler.py (hunk 2 above)
Comment (max-ipinfo) [reply ipinfo/handler.py]: Sorry for the confusion, @PriOliveira  🤦 !  I'll update the older lowercase convention in our code in a follow-up PR.

**Comment by max-ipinfo** [comment]
First off, sorry for my lack of context @PriOliveira.

 I am just starting with this repository and do not have a good understanding of the current conventions. Bear with me 😄 

> I added some tests to try to cover those cases, hope you don't mind.

I ran `pytest` and getting these errors (with version 3.12.4):
```
$ pytest
============================================================================================= short test summary info =============================================================================================
FAILED tests/handler_async_test.py::test_get_details_error[5xx_not_json] - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
FAILED tests/handler_async_test.py::test_get_details_error[4xx_json] - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
FAILED tests/handler_async_test.py::test_get_details_error[400] - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
FAILED tests/handler_async_test.py::test_get_details_quota_error - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
============================================================================== 4 failed, 22 passed, 22 skipped, 4 warnings in 1.36s ===============================================================================
```

Do these tests pass on your end? Could you share the module versions you are using?

I will most likely migrate to Poetry and upgrade dependency versions.

Thanks again!

**Comment by max-ipinfo** [comment]
I ran `pytest` at `HEAD` and getting the following errors:
```
============================================================================================= short test summary info =============================================================================================
FAILED tests/handler_async_test.py::test_get_batch_details[None] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details[1] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details[2] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details[3] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details_total_timeout[None] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details_total_timeout[1] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details_total_timeout[2] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
FAILED tests/handler_async_test.py::test_get_batch_details_total_timeout[3] - TypeError: Passing coroutines is forbidden, use tasks explicitly.
```

Sorry!

So feel free to only fix the following errors:
```
FAILED tests/handler_async_test.py::test_get_details_error[5xx_not_json] - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
FAILED tests/handler_async_test.py::test_get_details_error[4xx_json] - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
FAILED tests/handler_async_test.py::test_get_details_error[400] - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'
FAILED tests/handler_async_test.py::test_get_details_quota_error - AttributeError: 'MockResponse' object has no attribute 'wait_for_close'

```

**Comment by max-ipinfo** [comment]
I dug further and noticed that our CI pipeline and `requirements.txt` mention using v3.10.

After switching to v3.10 and running `pip install -r requirements.txt`, these failures are now warnings:
```
tests/default_cache_test.py ..                                                                                                                                 [  5%]
tests/details_test.py ....                                                                                                                                     [ 15%]
tests/handler_async_test.py ................                                                                                                                   [ 55%]
tests/handler_test.py ................                                                                                                                         [ 95%]
tests/init_test.py ..                                                                                                                                          [100%]

========================================================================== warnings summary ==========================================================================
tests/handler_async_test.py::test_get_batch_details[None]
tests/handler_async_test.py::test_get_batch_details[1]
tests/handler_async_test.py::test_get_batch_details[2]
tests/handler_async_test.py::test_get_batch_details[3]
tests/handler_async_test.py::test_get_batch_details_total_timeout[None]
tests/handler_async_test.py::test_get_batch_details_total_timeout[1]
tests/handler_async_test.py::test_get_batch_details_total_timeout[2]
tests/handler_async_test.py::test_get_batch_details_total_timeout[3]
  /Users/ipinfo/Desktop/python/ipinfo/handler_async.py:258: DeprecationWarning: The explicit passing of coroutine objects to asyncio.wait() is deprecated since Python 3.8, and scheduled for removal in Python 3.11.
    _, pending = await asyncio.wait(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================== 40 passed, 8 warnings in 6.13s ===================================================================
```

So let's test your changes with that same setup. Thank you!

**Review by max-ipinfo — APPROVED** [review]
Changes look good and tests pass with Python 3.10.

Thanks again @PriOliveira !

**Comment by PriOliveira** [comment]
I'm glad that in the end you figured out that it needs Python 3.10. 

Thanks a lot for the review, @max-ipinfo :)

**Pull request merged**